### ISSUE TITLE
Implement WebGPU async pipeline creation

### ISSUE THREAD

**Issue by gterzian** [servo/servo#31591]
Spec: https://gpuweb.github.io/gpuweb/#async-pipeline-creation
Stubs: https://github.com/servo/servo/pull/31194

How to do it: 

1. Store the promise on the device([here](https://github.com/servo/servo/pull/31194/files#diff-57b1e4d9cc2141b425d998d66541f5441c3e1ce4955c4713462e02132ba295e3R746)). Perhaps using a map of compute pipeline id to promise.  
2. Resolve the promise in [handle_server_msg](https://github.com/servo/servo/blob/3a5ca785d3ecc1fd6cb5a519cf1a91ac61e15c8c/components/script/dom/gpudevice.rs#L177)

**Comment by gterzian** [comment]
cc @sagudev

**Comment by sandeepB3** [comment]
Hey @gterzian, can I give this issue a try?

So elaborating on what you said, we are looking to create a mapping of `(id, promise)` inside `pub struct GPUDevice`, then we can use this map in `fn CreateComputePipelineAsync` to store the `promise` and `pipeline id`, once the mapping is done, we can send a request to the server to create the compute pipeline along with the `pipeline id`. 

Then the `fn handle_server_msg` can have a new variant `WebGPUOpResult::PipelineCreationResult` to handle results related to async pipeline creation, where the `mapping and pipeline id` is used to retrieve the promise and then it is either resolved or rejected. Is this what is expected?

**Comment by sandeepB3** [comment]
Hey @sagudev, are you still working on this draft: #31194 ?

**Comment by sagudev** [comment]
#31194 is now testing field for #31397, it's actually finished but it also includes CTS refresh and expectation stabilization (which actually works pretty well except for `webgpu:shader,execution,expression` tests that are still flaky).

Maybe it would make sense to land it just for partial CTS stabilization @gterzian?

**Comment by anamansari062** [comment]
@gterzian Can I work on this issue?

**Comment by gterzian** [comment]
@anamansari062 I think this one is doable as a first issue. Does the above description feel like it is giving you enough information? If not, I would look for an issue with a `B-newcomer` tag.

**Comment by anamansari062** [comment]
@gterzian It is doable. I have one question: what do you mean by store the promise? 
a. Store it in struct `GPUDevice`
b. Create a new struct for it

**Comment by gterzian** [comment]
@anamansari062 (also @sandeepB3 , sorry I missed your earleir message) I had another look and I think this is not a good first issue, because the `GPUDevice` is not plugged-in yet into the IPC messaging using [`AsyncWGPUListener`](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/gpu.rs#L47), which I thought it was via the `handle_server_msg` method(but it seems that method is only used to synchronously report errors, for example [here](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/gpudevice.rs#L379)). 

Also, doing this would require more than just storing the promise and later resolving it: it would likely also require storing additional info, and also making changes to the backend component at `component/webgpu` to send back additional info in the response message. What "additional info" means would require some looking into, I would say at least the "pipeline id" to match a given pending promise resolution pipeline to an incoming response to be handled. That additional data would then have to be part of the response, probably via a nested enum  at [`WebGPUOpResult`](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/webgpu/identity.rs#L24). 

So in other words, this would be quite a little project requiring some research and taking additional decisions based on what one finds.

**Comment by sagudev** [comment]
Ah, I think I got it, non async versions just returns pipeline objects (and in parallel runs actual creation in wgpu thread), async versions should just await for the creation to finish and then resolve promise, based on that. There is no need to store promise in device struct as this is abstracted via `response_async` that we should use. I will probably deal with it with new passes changes.
